torch_geometric/contrib/flag/flag.py

```diff
@@ -0,0 +1,249 @@
+            y_true (torch.Tensor): Graph-level, node-level, or edge-level
+                ground-truth labels.
+            train_idx (torch.Tensor): Indices of training data.
+            step_size (float): The step size to take during the
+                gradient ascent on the perturbation. It is also used for
+                the ranges when initializing the perturbation. If
+                `step_size_unlabeled` is set, `step_size` will only be applied
+                to the labeled nodes.
+            step_size_unlabeled (Optional[float]): The step size to take
+                during the gradient ascent on the perturbation for unlabeled
+                nodes. This is also used for the ranges when initializing the
+                perturbation. If this is not set, `step_size` is applied to
+                both labeled and unlabeled nodes.
+            n_ascent_steps (int): The number of forward and backward passes
+                to take. Note that the FLAG paper uses 3 for experiments.
+
+        :returns:
+            - self.loss (torch.nn.Module) - The loss after the final \
+                iteration of the FLAG algorithm.
+            - out (torch.Tensor) - The model output on the perturbed inputs \
+                from the last iteration of the FLAG algorithm.
+        """
+
+        # Check the parameters are specified correctly
+        if not isinstance(n_ascent_steps, int) or n_ascent_steps <= 0:
+            raise ValueError(f"Invalid n_ascent_steps: {n_ascent_steps}." +
+                             " n_ascent_steps should be a positive integer.")
+
+        if not isinstance(step_size, numbers.Number) or step_size <= 0:
+            raise ValueError(f"Invalid step_size: {step_size}." +
+                             " step_size should be a positive float.")
+
+        if step_size_unlabeled is not None and (
+                not isinstance(step_size_unlabeled, numbers.Number)
+                or step_size_unlabeled <= 0):
+            raise ValueError(
+                f"Invalid step_size_unlabeled: {step_size_unlabeled}." +
+                " step_size_unlabeled should be None or a positive float.")
+
+        # Code below adapted from:
+        #   https://github.com/devnkong/FLAG/blob/main/deep_gcns_torch
+        #     /examples/ogb/ogbn_arxiv/main.py#L56
+        #   and
+        #   https://github.com/devnkong/FLAG/blob/main/deep_gcns_torch
+        #     /examples/ogb/attacks.py
+
+        training_labels = y_true.squeeze(1)[train_idx]
+        unlabeled_idx = torch.tensor(
+            list(set(range(x.shape[0])) - set(train_idx)), dtype=torch.long)
+        self._set_perturb(x, train_idx, unlabeled_idx, step_size,
+                          step_size_unlabeled)
+
+        self.model.train()
+        self.optimizer.zero_grad()
+
+        [c.on_ascent_step_begin(0, None) for c in self.callbacks]
+
+        # Perturb the input and calculate the loss
+        out = self.model(x + self.perturb, edge_index)[train_idx]
```

Comment:
@EdisonLeeeee instead of explicitly adding those arguments, would it make sense for `flag.forward` to take in `**kwargs` which get passed to the model's `forward` function? 

Also, just for my own understanding, what is the `batch` argument used for in a model's forward function?